documentation/projects/proposals/trust_and_safety/detecting_sensitive_textual_content/20240903-implementation_plan_undo_split_filtered_indices.md

```diff
@@ -0,0 +1,830 @@
+approach. The API will use a feature flag to determine which querying approach
+to use. This makes it easy for us to switch to the old querying approach in case
+of failure, without needing to roll back or revert code, which could impact
+other ongoing projects.
+
+Finally, once we have confidence the new approach is working, we begin the phase
+of cleaning up the multi-index approach. There are two major parts to this:
+removing the filtered index management from the data refresh, and removing the
+API code that utilises it. Neither of these appear to have much complexity. Both
+are described in detail in the [step-by-step section](#step-by-step-plan).
+
+### Evaluating the performance of the new approach
+
+We should make an effort to compare the new approach to the old. There are two
+methods we can use, one which seems intuitive but that I believe will not work,
+and another which I believe will work, and which I propose we follow.
+
+To attempt a simultaneous comparison of the two strategies, we could do the
+following: preserve the ability for both strategies in the API, and direct a
+portion of search requests to the new method. Using a new, temporary response
+header, expose the search method used for the request in the Nginx logs. That
+will enable us to query Logs Insights and determine whether there is a
+meaningful difference in the average, median, p95, and p99 search timings for
+the two methods over the same time period. This is an "ad-hoc" approach to
+compensate
+[for the lack of an existing A/B experimentation framework](https://github.com/WordPress/openverse/issues/421).
+
+However, this presents a problem: it does not compare the final outcome of this
+plan with the current state of things. The motivation for removing the filtered
+index is predicated on the idea that moving to a single index will improve
+Elasticsearch's ability to utilise its query cache and reduce disk operations.
+If we are simultaneously using the single and multi-index approaches, the
+behaviour of the cluster and application will not reflect the reduction in disk
+operations or increase in cache utilisation. That is because the thing causing
+the current behaviour will still be operating. In fact, trying both
+simultaneously would probably make things worse for all queries across the
+board. We would be sending rather different types of queries, still to multiple
+indices, thereby reducing Elasticsearch's opportunity for cache utilisation and
+thus increasing its disk operations.
+
+To actually implement a simultaneous comparison, I hypothesise that we would
+need to have identically resourced clusters, one for executing the new type of
+query, and one for the old type. I believe because the performance question in
+this plan has to do with reducing disk operations and increasing cache
+utilisation, that the metrics are unique from relevance optimisations, which we
+would be able to measure side by side.
+
+While possible in theory, it is too complex!
```

Comment:
One approach that we could take is to qualitatively estimate if the single index approach makes things better is to vary the portion of queries that use the new index. 

If our hypothesis is correct, increasing the portion of queries using the new index should correspond to a in overall resource usage. This way we also don't have to drop the feature flag outright and we can gradually switch over 100% of requests to the new index.

Reply:
True, it would be cool if we had methods for doing that! We don't have anything currently set up that let us adjust settings on the fly like that. It would be interesting if we had a Django model for settings like that where we might want more responsive control than environment variables provide. I anticipate something like that would come from #421... I'd be wary of implementing anything like that for this, and doing several deployments with small environment variable changes each time would be a headache compared to just doing a clean switch over and rolling back if needed.

> If our hypothesis is correct, increasing the portion of queries using the new index should correspond to a in overall resource usage

One part of this that I'm not confident in is... how many queries would need to be redirect to the new method for it to make a difference for Elasticsearch? The resource consumption comes from the fact that ES has to pull results from two queries instead of one (and sometimes those raw documents are identical, so it's wasting cache at least some of the time).

In this particular case, it's such a radically different approach between the two query methods (one index vs two) that I don't know what to anticipate from Elasticsearch if it were getting both. Other than that I'm fairly confident it would worsen caching because it would be getting more "types" of queries to the unfiltered index. If at some point of switching to more "new" queries it ever does improve cacheability, the percentage of "old" style queries would have to be very low, I would think. How ES responds to the frequency, and whether we would start to see improvements in cache performance over time, are a question. My guess is never, so long as both indices are being used, it will continue to require pulling from both indices on disk, and end up using a lot of its cache to do so.

That is really only a guess. My point is that I just don't know enough about how ES caches to say that we would see a corresponding change by moving the search method over slowly.

---

That being said, it would be cool if we had more responsive controls and things like what you're saying were more easily possible. I could see it being very useful for performance testing new query functions, for example. If we could ramp up, rather than cut over immediately, and not need to have a bunch of individual deployments for env variable changes, we'd probably get a better picture of how things compare under certain conditions. I head someone from Instagram talk about doing this with traffic to Instagram deployments running different modifications of CPython (part of their fast-python stuff), where they slowly move a percentage of traffic over and get a read on resource usage. Very cool! Must be lots of interesting dev tooling to make it nice to do on a regular basis.